src/coins.cpp

```diff
@@ -260,6 +295,240 @@ void CCoinsViewCache::BringBestAnchorIntoCache(
+            if (peak_pos <= treeLength) {
+                draftMMRNode(entry_indices, entries, GetHistoryAt(epochId, peak_pos-1), h, peak_pos);
+
+                last_peak_pos = peak_pos;
+                last_peak_h = h;
+
+                // right sibling
+                peak_pos = peak_pos + (1 << (h + 1)) - 1;
+            }
+        }
+    }
+
+    total_peaks = entries.size();
+
+    // early return if we don't extra nodes
+    if (!extra) return total_peaks;
+
+    h = last_peak_h;
+    peak_pos = last_peak_pos;
+
+
+    //           P
+    //          / \
+    //         /   \
+    //        / \   \
+    //       /    \  \
+    //    _A_      \   \
+    //   /   \_     B   \
+    //  / \   / \   / \  C
+    // /\ /\ /\ /\ /\ /\ /\
+    //                   D E
+    //
+    //
+    // For extra peaks needed for deletion, we do extra pass on right slope of the last peak
+    // and add those nodes + their siblings. Extra would be (D, E) for the picture above.
+    while (h > 0) {
+        uint32_t left_pos = peak_pos - (1<<h);
+        uint32_t right_pos = peak_pos - 1;
+        h = h - 1;
+
+        // drafting left child
+        draftMMRNode(entry_indices, entries, GetHistoryAt(epochId, left_pos-1), h, left_pos);
+
+        // drafting right child
+        draftMMRNode(entry_indices, entries, GetHistoryAt(epochId, right_pos-1), h, right_pos);
+
+        // continuing on right slope
+        peak_pos = right_pos;
+    }
+
+    return total_peaks;
+}
+
+HistoryCache& CCoinsViewCache::SelectHistoryCache(uint32_t epochId) const {
+    auto entry = historyCacheMap.find(epochId);
+
+    if (entry == historyCacheMap.end()) {
+        historyCacheMap[epochId] = HistoryCache(
+            base->GetHistoryLength(epochId),
+            base->GetHistoryRoot(epochId),
+            epochId
+        );
+
+        return historyCacheMap.find(epochId)->second;
+    }
+
+    return entry->second;
+}
+
+void CCoinsViewCache::PushHistoryNode(uint32_t epochId, const HistoryNode node) {
+    auto treeLength = GetHistoryLength(epochId);
+    HistoryCache& historyCache = SelectHistoryCache(epochId);
+
+    if (treeLength == 0) {
+        // special case, it just goes into the cache right away
+        historyCache.Extend(node);
+
+        if (librustzcash_mmr_hash_node(epochId, node.data(), historyCache.root.begin()) != 0) {
+            throw std::runtime_error("hashing node failed");
+        };
+
+        return;
+    }
+
+    std::vector<HistoryEntry> entries;
+    std::vector<uint32_t> entry_indices;
+
+    PreloadHistoryTree(epochId, false, entries, entry_indices);
+
+    uint256 newRoot;
+    std::array<HistoryNode, 32> appendBuf;
+
+    uint32_t appends = librustzcash_mmr_append(
+        epochId, 
+        treeLength,
+        entry_indices.data(),
+        entries.data()->data(),
+        entry_indices.size(),
+        node.data(),
+        newRoot.begin(),
+        appendBuf.data()->data()
+    );
```

Comment:
I was discussing [my comment here](https://github.com/zcash/zcash/pull/4264#discussion_r393442120) with @daira in our pairing, and ze was not sure whether this was valid, or whether there might be padding introduced. It turns out that [there is no guarantee that `sizeof(std::array<unsigned char, N>) == N`](https://stackoverflow.com/questions/25173462/what-is-the-sizeof-stdarraychar-n), so it is incorrect to pass a pointer to `std::vector<std::array<unsigned char, N>>` (specifically `entries.data()->data()`) through the C API, and treat it as a slice of fixed-length byte arrays on the Rust side. Instead, we need to either use a C struct (with an equivalent `#[repr(transparent)]` struct on the Rust side), or use C arrays instead of `std::array`.

Reply:
Nice catch, thanks

Reply:
Did we add `#[repr(transparent)]` and/or make size and alignment assertions on the Rust side corresponding to the `static_assert`s on the C++ side? (I don't entirely trust the ABIs to match up otherwise.)

Reply:
On the Rust side we are treating it as a slice of byte arrays, which in Rust are [guaranteed to be contiguous](https://rust-lang.github.io/unsafe-code-guidelines/layout/arrays-and-slices.html):

> The alignment of array types is greater or equal to the alignment of its element type. If the element type is `repr(C)` the layout of the array is guaranteed to be the same as the layout of a C array with the same element type.

> When the element _size_ is a multiple of the element's _alignment_, then `stride == size`, and the elements are laid out contiguously in memory, e.g., `[u8; 4]`. In this case, the size of the array can be computed as `size_of::<T>() * N`, and a pointer to the `i`-th element of the array can be obtained by offsetting a pointer to the first element of the array by `i`.

> The layout of a slice `[T]` of length `N` is the same as that of a `[T; N]` array.